util/apiwarnings/handler.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiwarnings
+
+import (
+	"regexp"
+	"sync"
+
+	"github.com/go-logr/logr"
+)
+
+// DiscardMatchingHandlerOptions configures the handler created with
+// NewDiscardMatchingHandler.
+type DiscardMatchingHandlerOptions struct {
+	// Deduplicate indicates a given warning message should only be written once.
+	// Setting this to true in a long-running process handling many warnings can
+	// result in increased memory use.
+	Deduplicate bool
+
+	// Expressions is a slice of regular expressions used to discard warnings.
+	// If the warning message matches any expression, it is not logged.
+	Expressions []regexp.Regexp
+}
+
+// NewDiscardMatchingHandler initializes and returns a new DiscardMatchingHandler.
+func NewDiscardMatchingHandler(l logr.Logger, opts DiscardMatchingHandlerOptions) *DiscardMatchingHandler {
+	h := &DiscardMatchingHandler{logger: l, opts: opts}
+	if opts.Deduplicate {
+		h.logged = map[string]struct{}{}
+	}
+	return h
+}
+
+// DiscardMatchingHandler is a handler that discards API server warnings
+// whose message matches any user-defined regular expressions.
+type DiscardMatchingHandler struct {
+	// logger is used to log responses with the warning header
+	logger logr.Logger
+	// opts contain options controlling warning output
+	opts DiscardMatchingHandlerOptions
+	// loggedLock guards logged
+	loggedLock sync.Mutex
+	// used to keep track of already logged messages
+	// and help in de-duplication.
+	logged map[string]struct{}
+}
+
+// HandleWarningHeader handles logging for responses from API server that are
+// warnings with code being 299 and uses a logr.Logger for its logging purposes.
+func (h *DiscardMatchingHandler) HandleWarningHeader(code int, _, message string) {
+	if code != 299 || message == "" {
+		return
+	}
+
+	for _, exp := range h.opts.Expressions {
+		if exp.MatchString(message) {
+			return
+		}
+	}
+
+	if h.opts.Deduplicate {
+		h.loggedLock.Lock()
+		defer h.loggedLock.Unlock()
+
+		if _, alreadyLogged := h.logged[message]; alreadyLogged {
+			return
+		}
+		h.logged[message] = struct{}{}
```

Comment:
Hm, these entries are never removed, so this is basically a memory leak, right? :)

If I see correctly that's basically just the same feature as in the CR KubeAPIWarningLogger. I'm just wondering if it's a useful feature to have. 

I mean how important is it to deduplicate warnings :)

Reply:
> Hm, these entries are never removed, so this is basically a memory leak, right? :)

Good point! I should have questioned the code more closely.

The de-duplication comes from [KubeAPIWarningLogger](https://github.com/kubernetes-sigs/controller-runtime/blob/b8db76e66383819c94f530bd4b8dea60025b716f/pkg/log/warning_handler.go#L50-L51), and that comes from client-go's "ready to use" (but not used by default) [warning handler](https://github.com/kubernetes/client-go/blob/master/rest/warnings.go#L108).

> I mean how important is it to deduplicate warnings :)

It could be useful to de-duplicate warnings within some time window. For example, to avoid 10 identical warnings logged in within the last 60 seconds. The current implementation has an effectively infinite time window :grimacing:.

Reply:
I've removed deduplication.

Reply:
As an aside, what do you think about disabling deduplication by default in controller-runtime? We currently [enable deduplication](https://github.com/kubernetes-sigs/controller-runtime/blob/4a356a8b1a22aafe55d07d067e935dc64fb90919/pkg/client/client.go#L120) by default.

Reply:
Sounds good thx! Also agree that it should be disabled in CR